File: leetcode/data-structures-and-algorithms/heaps/1046-last-stone-weight.java

// https://leetcode.com/problems/last-stone-weight/description/

import java.util.Collections;
import java.util.PriorityQueue;

class Solution {
    public int lastStoneWeight(int[] stones) {
        // max heap
        PriorityQueue<Integer> heap = new PriorityQueue<>(stones.length, Collections.reverseOrder());
        for (var s : stones) {
            heap.add(s);
        }

        while (heap.size() > 1) {
            var s1 = heap.poll();
            var s2 = heap.poll();

            // not equal
            if (s1 != s2) {
                // always > 0 because s1 is bigger by max heap definition
                heap.add(s1 - s2);
            }
        }

        if (heap.size() > 0) {
            return heap.peek();
        }

        return 0;
    }
}

File: leetcode/data-structures-and-algorithms/heaps/1046-last-stone-weight.ts
// https://leetcode.com/problems/last-stone-weight/description/

import { MaxPriorityQueue } from "@datastructures-js/priority-queue";

function lastStoneWeight(stones: number[]): number {
  const heap = new MaxPriorityQueue<number>();
  for (const s of stones) {
    heap.push(s);
  }

  while (heap.size() > 1) {
    const y = heap.pop();
    const x = heap.pop();

    if (y !== x) {
      heap.push(y - x);
    }
  }

  if (heap.size() > 0) {
    return heap.front();
  }

  return 0;
}
